README.md
# FOLIO Order import tool

Workaround for importing orders from MARC records until FOLIO Data Import supports importing MARC records to create
orders.

## What does it do?

* It takes an uploaded file that contains MARC records, and creates orders, instances, holdings and items in FOLIO Orders and FOLIO Inventory.
* It uses the 980 field to get fund code, vendor code, price, notes, electronic/print indicator, etc. to create the order.
* It lets FOLIO Orders create the instance, holdings and item records using the "createInventory" value in the order line
* It only asks FOLIO Orders to create items if the material is print
* It uses a property file to determine constants such as locations, fiscal year, note type, and material type and default text for electronic resources

For details, see [What data are populated to FOLIO](#what-data-are-populated-to-folio).

### It has two modes, 'validate only' or 'validate and import'

The MARC file can be run through validation without actually importing anything to FOLIO. Choose a MARC file in the tool's UI and click `Analyze MARC records`. For details about the checks it performs, see [Validation of incoming MARC records](#validation-of-incoming-marc-records)

When running an actual import, the exact same checks are performed for each record, and the tool then acts on the results according to the setting of configuration parameter `onValidationErrors`. See options for this setting in the [configuration table](#how-to-configure-the-service) 

![ui-screen-shot](https://user-images.githubusercontent.com/11644885/146233019-fe7b87d0-a3a0-4372-b135-6a22d14c06e5.png)


Imports can take a while for files with many MARC records, so imports are started in the back-ground. Right after starting an import, the UI will display a status with the state 'started' and a count of MARC records in the file and a 'Refresh' link that you can click to get updates on progress. 

![ui-import-started](https://user-images.githubusercontent.com/11644885/146233538-9b19b43f-2875-491b-b41f-68631bcc66fd.png)

Once the job is finished, clicking the 'Refresh' will display a state of 'done' and detailed results for each MARC record in the file.

## If you want to try it

* Clone the repo.
* The tool expects a properties file to be specified by one of these methods:
    1. A environment variable on the command line, named 'config', like `-Dconfig=/path/to/your.properties`.
    1. A ServletContext attribute, named 'config', specified in the servlet container configuration. Useful for [deploying multiple instances](#multiple-instances) in one servlet container.
    1. If neither of the above 'config' variables are set, it expects to find the properties file here: `/yourhomefolder/order/import.properties`.
* You will have to set your Okapi address, FOLIO tenant, userid and password in the properties, and you may have to adjust the file upload path where it will save
  the uploaded file and store a history of validations and imports, see the [How to configure the service](#how-to-configure-the-service)
* Call: `mvn jetty:run [-Dconfig=path-to-properties-file]`
* If the configuration is okay, it should start a jetty server, and you should be able to point your browser to http://localhost:8888/import and try
  it
* However, unless `exitOnConfigErrors` is set to false in the properties, the service will stop if it detects fatal configuration problems
* Unless `exitOnAccessErrors` is set to false, the service will stop if it fails to log in to Okapi. 
* And, unless `exitOnFailedIdLookups` is set to false, the service will stop if it could not find FOLIO UUIDs for names or codes
  defined in the properties.
* We've included example MARC files, but you will have to update them with your vendor, fund, object codes, etc.
* To effectuate changes of import properties, restart the service

### Docker image

You can build a Docker image using the [Dockerfile](Dockerfile) in this repository.

1. Build the WAR for the webapp: `mvn install`
1. Build the Docker image: `docker build .`
1. Run the container: `docker run -d -p 8080:8080 <imageId>`

This will run a [Jetty](https://hub.docker.com/_/jetty) container with the order import webapp as the root, using the
default configuration in the [import.properties](import.properties) file. This will work against the
FOLIO [folio-snapshot](https://folio-snapshot.dev.folio.org) reference environment.

To override the default configuration, mount your configuration to `/var/lib/jetty/order/import.properties` on the
container e.g.:

    docker run -d -v $(pwd)/order:/var/lib/jetty/order -p 8080:8080 <imageId>

### Multiple instances

You may want to run multiple instances of the tool in the same server environment, using separate configuration files. For example, this would allow separate production and test instances that each point to the appropriate FOLIO API endpoint.

Note that the webapp expects to be running at the root path of its server.  So first set up separate virtual hosts for each instance.

* Add a CNAME my-instance-name.domain.edu referencing the actual server.
* Configure a reverse proxy to point the root path of that virtual host to a distinct path on the actual server.  For example in Apache, 

```
<VirtualHost ....>
  ....
  ServerName my-instance-name.domain.edu
  
  ProxyPreserveHost On
  
  RewriteEngine on
  RewriteRule ^$ /import [L]
  
  <Location />
    ....
   
    RequestHeader set X-Forwarded-Proto "https" env=HTTPS
    ProxyPass http://my-instance-name.domain.edu:8080/my-instance-name/
    ProxyPassReverse http://my-instance-name.domain.edu:8080/my-instance-name/
  </Location>
  
  ....
</VirtualHost>
```
Then configure each app instance to point to a separate properties file using a ServletContext attribute in its servlet configuration.  In Jetty, as an example, define a servlet properties file (in the same directory and with the same filename as the .war, but extension .xml) such as:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">

<Configure class="org.eclipse.jetty.webapp.WebAppContext">
  <Set name="contextPath">/my-instance-name</Set>
  <Set name="war">/usr/share/jetty9/webapps/my-instance-name.war</Set>
  <Get name="ServletContext">
    <Call name="setAttribute">
      <Arg>config</Arg>
      <Arg>/some-folder-with-configuration-files/import-my-instance-name.properties</Arg>
    </Call>
  </Get>
</Configure>
```

## How to configure the service

The startup configuration has parameters controlling FOLIO access, start-up behavior, server settings, process controls, UI settings and static values populated to FOLIO. The table show all parameters except for the static FOLIO data that are described in the subsequent paragraphs.

***FOLIO access settings, startup behavior, server settings for the tool, import process controls, UI settings***

| Name                      | Values                                                                                                                                 | Default                      | What it does                                                                                                                                                                                                                                                                                                          |
|---------------------------|----------------------------------------------------------------------------------------------------------------------------------------|------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **FOLIO access**          ||||
| baseOkapiEndpoint         | Protocol and domain of the FOLIO backend REST service<br/>i.e. https://folio-snapshot.dev.folio.org/                                   ||                              |
| tenant                    | FOLIO tenant, i.e. 'diku'                                                                                                              |||
| okapi_username            | The FOLIO user name of the import user i.e. 'diku_admin'                                                                               |||
| okapi_password            | The password for the import user                                                                                                       |||
| **Startup**               ||||
| exitOnConfigurationErrors | boolean ¹                                                                                                                              | true                         | If set to true (the default), the service will exit if it detects fatal problems with the configuration properties -- like missing mandatory properties. <br/>If set to false, it will merely log configuration problems.                                                                                             | 
| exitOnAccessErrors        | boolean ¹                                                                                                                              | true                         | If set to true (the default), the service will exit if it fails to gain access to FOLIO for any reason (wrong URL, tenant, username or service down etc). <br/>If set to false, it will merely log access problems.                                                                                                   |
| exitOnFailedIdLookups     | boolean ¹                                                                                                                              | true                         | If set to true (the default), the service will exit if it fails to resolve configured names of codes to FOLIO UUIDs. <br/>If set to false, it will merely log missing values.                                                                                                                                         |
| **Server**                ||||
| uploadFilePath            | File system path. The parent directory must exist on the server. The given subdirectory will be created if it doesn't exist already. ² | /var/tmp/folio-order-import/ | Directory for storing uploaded MARC files and the outcomes of imports.                                                                                                                                                                                                                                                |
 | daysToKeepResults         | number of days ³                                                                                                                       | 365                          | Results log files older than this number of days will be deleted from the server.                                                                                                                                                                                                                                     |
| marcMapping               | `chi`, `lambda`, or `sigma`                                                                                                            | chi                          | Selects a MARC mapping option. The available mapping options are described in the mapping tables below.                                                                                                                                                                                                               |
| **Processing**            ||||
| purchaseOrderUnit         | `file` or `record`                                                                                                                     | record                       | Instructs the tool to create one purchase order per file or one purchase order per record (default).                                                                                                                                                                                                                  |
| onValidationErrors        | `cancelAll`, `skipFailed` or `attemptImport`                                                                                           | cancelAll                    | If one or more records fail the initial validation check, this setting will cause the service to either cancel the entire import (`cancelAll`), skip the current, failed record (`skipFailed`), or attempt import anyway (`attemptImport`). With the last option, the import itself would presumably eventually fail. |
| onIsbnInvalid             | `removeIsbn`, `reportError` or `doNothing`                                                                                             | reportError                  | Controls if the tool should report error and perhaps skip the record, or remove the ISBN to ingest, or do nothing (which should cause the import to error out later)                                                                                                                                                  | 
| **Tool UI**               |||
| daysToShowResults         | number of days ³                                                                                                                       | 14                           | Results are listed in the UI for this number of days after first created. After that they will be skipped (but not deleted)                                                                                                                                                                                           |
| tzTimeZone                | A [tz time zone](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones), ie `America/Chicago`                                   | Europe/Stockholm             | Sets the time zone of dates in the UI                                                                                                                                                                                                                                                                                 |
| locale                    | Language and country, i.e. `en-US`                                                                                                     | sv-SE                        | Formats dates in the UI                                                                                                                                                                                                                                                                                               | 
| folioUiUrl                | Protocol and domain of FOLIO UI, ie https://folio-snapshot.dev.folio.org/ ²                                                            | none                         | If provided, links to FOLIO's UI will be displayed for records in the import log.                                                                                                                                                                                                                                     |
| folioUiInventoryPath      | Path to the Inventory UI ²                                                                                                             | inventory/view               | Used for a link in the import log to the Instance in UI Inventory.                                                                                                                                                                                                                                                    |
| folioUiOrdersPath         | Path to the Orders UI ²                                                                                                                | orders/view                  | Used for a link in the import log to the order in UI Orders.                                                                                                                                                                                                                                                          |


1) Boolean settings: `true`, `TRUE`, `yes`, `YES`, `y`, `Y`, and `1` will resolve to **true**, while `false`, `FALSE`, `no`, `NO`, `n`, `N`
, and `0` will resolve to **false**
2) file paths, URLs and API paths can be with our without the ending slash (`/`)
3) if the provided config value is not a valid number, the default will apply

## What data are populated to FOLIO
The tool populates FOLIO Orders and Inventory with data from three different sources: 
1) The incoming MARC records
2) Static values defined as configuration parameters
3) Static values hard-coded in the program. 

### Data from the incoming MARC records
The tool provides three, slightly different sets of MARC mappings. All three share a basic set of mappings but then extend that with some mappings of their own. 

The desired mapping is selected at startup by setting the parameter `marcMapping` to either `chi` (default), `lambda` or `sigma`. 

#### Core mapping table

| MARC_fields¹       | Description                    | Target properties                                                                                                                                                                                     | Required                                 | Default                                                                                | Content (incoming)                                                                                     |
|--------------------|--------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------|----------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------|
| 020 $a ($c $q)     | ISBN                           | orderLine.details.productIds[] <br/>instance.identifiers[]                                                                                                                                            | No, but a note will be logged if missing |                                                                                        |

NEWS.md
## 1.9.0 not released

* Display OIT's POM version on configuration view page

## 1.8.1 2023-03-07

* Use exact match for title look-up (SUP-50607)

## 1.8.0 2023-03-07

* Switch Instance lookup from Inventory to Search (SUP-50607)

## 1.7.3 2022-09-09

* Chi mapping: Restore note type to electronic bookplate for items.
* Set jetty artifact versions to `9.4.48.v20220622` and pin them.

## 1.7.2 2022-06-09

* Update Dockerfile to specify Jetty 10 as the base image.

## 1.7.1 2022-06-09

* Sigma mapping: Only sets note type on items if there's a donor. 

## 1.7.0 2022-05-01

* Sets instance.indexTitle to title with initial article removed.

## 1.6.0 2022-05-01

* Escapes all title characters that are CQL special characters.
* Maps name-of-part (245$p) to end of title.
* Maps editions from first 250$a.
* Maps series from first 490, all subfields except $y, $z.
* Maps languages from 041$a.
* Sets holdings type for physical resources, too (as for electronic).
* Maps resource type from first 336$a (default still 'text').
* Maps instance formats from first 337$a " -- " 338$a, concatenated.
* Chi mapping: Sets item copy number to 'c.1'.

## 1.5.1 2022-03-17

* Workaround for CQL parser issue affecting titles with quotes  
* Log4J upgrade for security issue

## 1.5.0 2022-02-02

* Supports ISBNs with last digit of 'X'.
* Lambda mapping: ignores 856$x 'userLimit' field, so it can be used with non-integer values.
* Supports passing import properties file name as ServletContext attribute.
* Chi mapping: Add OCLC number to Instance identifiers.
* Chi mapping: Use only ISBN for product ID (and thus for linking to existing Instance).
* Update only newly created item(s); in the rare case holdings record has previously existing items.

## 1.4.2 2022-01-21

* Fixes premature use of coming 'holdingId' property in JSON response to POST of PO

## 1.4.1 2022-01-20

* Fixes bug in location validation for MARC records with multiple 980s

## 1.4.0 2022-01-19

* Supports one-PO-per-file and multiple-locations-per-MARC-record
* Adds MARC mapping extension 'sigma'

## 1.3.2 2022-01-18

* Bug-fix: Validates Instance ISBNs without including any qualifiers.

## 1.3.1 2021-12-17

* Bug-fix: Removes mistaken attempt to update Item for electronic

## 1.3.0 2021-12-13

* Chi mapping: maps barcode from 980$o
* Provides three sets of MARC-to-FOLIO mappings: Chi, Lambda, Sigma.
* The tool adapts to coming schema change in Orders wrt acquisition method
* The validation logic checks fund, budget, budget expense class, and price.

## 1.2.0 2021-12-09

* Configurable time zone for dates shown in UI
* Configurable locale for dates
* Redundant request to Orders is removed
* Empty results file stored immediately, to avoid exceptions on early "Refresh" in UI

## 1.1.0 2021-12-06

* Runs import jobs asynchronously and provides UI mechanisms to follow progress
* Stores import JSON logs at the file upload path, UI provides listing of import history
* Properly sets default materialType if not provided in properties
* Allows all paths in properties to be with our without the ending slash

## 1.0.0 2021-12-02


src/main/java/org/olf/folio/order/Config.java
package org.olf.folio.order;

import java.io.IOException;
import org.marc4j.marc.Record;
import org.olf.folio.order.mapping.MarcToFolio;
import org.olf.folio.order.mapping.MarcMapChi;
import org.olf.folio.order.mapping.MarcMapLambda;
import org.olf.folio.order.mapping.MarcMapSigma;

import javax.servlet.ServletContext;
import java.time.ZoneId;
import java.time.zone.ZoneRulesException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import org.olf.folio.order.utils.PropertiesReader;

public class Config {

  // FOLIO ACCESS
  public static final String P_BASE_OKAPI_ENDPOINT = "baseOkapiEndpoint";
  public static final String P_BASE_OKAP_ENDPOINT = "baseOkapEndpoint"; // previous spelling
  public static final String P_TENANT = "tenant";
  public static final String P_OKAPI_USERNAME = "okapi_username";
  public static final String P_OKAPI_PASSWORD = "okapi_password";
  // OPERATIONS
  public static final String P_UPLOAD_FILE_PATH = "uploadFilePath";
  public static final String P_DAYS_TO_KEEP_RESULTS = "daysToKeepResults";
  // UI
  public static final String P_DAYS_TO_SHOW_RESULTS = "daysToShowResults";
  public static final String P_TZ_TIME_ZONE = "tzTimeZone";
  public static final String P_LOCALE = "locale";
  // STATIC VALUES
  public static final String P_FISCAL_YEAR_CODE = "fiscalYearCode";
  public static final String P_NOTE_TYPE = "noteType";
  public static final String P_MATERIAL_TYPE = "materialType";
  public static final String P_PERM_LOCATION = "permLocation";
  public static final String P_PERM_E_LOCATION = "permELocation";
  public static final String P_PERM_LOCATION_WITH_INVOICE_IMPORT = "permLocationWithInvoiceImport";
  public static final String P_PERM_E_LOCATION_WITH_INVOICE_IMPORT = "permELocationWithInvoiceImport";
  public static final String P_TEXT_FOR_ELECTRONIC_RESOURCES = "textForElectronicResources";
  public static final String P_PAYMENT_METHOD = "paymentMethod";
  // MARC MAPPING
  public static final String P_MARC_MAPPING = "marcMapping";
  public static final String V_MARC_MAPPING_CHI = "chi";
  public static final String V_MARC_MAPPING_LAMBDA = "lambda";
  public static final String V_MARC_MAPPING_SIGMA = "sigma";
  // PROCESSING INSTRUCTIONS
  public static final String P_PURCHASE_ORDER_UNIT = "purchaseOrderUnit";
  public static final String V_PURCHASE_ORDER_UNIT_FILE = "file";
  public static final String V_PURCHASE_ORDER_UNIT_RECORD = "record";
  public static final String P_IMPORT_INVOICE = "importInvoice";
  public static final String P_FAIL_IF_NO_INVOICE_DATA = "failIfNoInvoiceData";
  public static final String P_EXIT_ON_CONFIG_ERRORS = "exitOnConfigErrors";
  public static final String P_EXIT_ON_ACCESS_ERRORS = "exitOnAccessErrors";
  public static final String P_EXIT_ON_FAILED_ID_LOOKUPS = "exitOnFailedIdLookups";
  public static final String P_ON_VALIDATION_ERRORS = "onValidationErrors";
  public static final String V_ON_VALIDATION_ERRORS_CANCEL_ALL = "cancelAll";
  public static final String V_ON_VALIDATION_ERRORS_SKIP_FAILED = "skipFailed";
  public static final String V_ON_VALIDATION_ERRORS_ATTEMPT_IMPORT = "attemptImport";
  public static final String P_ON_ISBN_INVALID = "onIsbnInvalid";
  public static final String V_ON_ISBN_INVALID_REPORT_ERROR = "reportError";
  public static final String V_ON_ISBN_INVALID_REMOVE_ISBN = "removeIsbn";
  public static final String V_ON_ISBN_INVALID_DO_NOTHING = "doNothing";
  // OPTIONAL FEEDBACK SETTINGS
  public static final String P_FOLIO_UI_URL = "folioUiUrl";
  public static final String P_FOLIO_UI_INVENTORY_PATH = "folioUiInventoryPath";
  public static final String P_FOLIO_UI_ORDERS_PATH = "folioUiOrdersPath";

  public static final List<String> KNOWN_PROPERTIES = Arrays.asList(
          P_BASE_OKAP_ENDPOINT,
          P_BASE_OKAPI_ENDPOINT, P_DAYS_TO_SHOW_RESULTS,
          P_DAYS_TO_KEEP_RESULTS,
          P_EXIT_ON_ACCESS_ERRORS,
          P_EXIT_ON_CONFIG_ERRORS,
          P_EXIT_ON_FAILED_ID_LOOKUPS,
          P_FAIL_IF_NO_INVOICE_DATA,
          P_FISCAL_YEAR_CODE,
          P_FOLIO_UI_INVENTORY_PATH,
          P_FOLIO_UI_ORDERS_PATH,
          P_FOLIO_UI_URL,
          P_IMPORT_INVOICE,
          P_LOCALE,
          P_MARC_MAPPING,
          P_MATERIAL_TYPE,
          P_NOTE_TYPE,
          P_OKAPI_PASSWORD,
          P_OKAPI_USERNAME,
          P_ON_ISBN_INVALID,
          P_ON_VALIDATION_ERRORS,
          P_PAYMENT_METHOD,
          P_PERM_E_LOCATION,
          P_PERM_LOCATION,
          P_PERM_E_LOCATION_WITH_INVOICE_IMPORT,
          P_PERM_LOCATION_WITH_INVOICE_IMPORT,
          P_PURCHASE_ORDER_UNIT,
          P_TENANT,
          P_TEXT_FOR_ELECTRONIC_RESOURCES,
          P_TZ_TIME_ZONE,
          P_UPLOAD_FILE_PATH
  );

  public static String baseOkapiEndpoint;
  public static String folioUiUrl;
  public static String folioUiInventoryPath;
  public static String folioUiOrdersPath;
  public static String apiUsername;
  public static String apiPassword;
  public static String tenant;
  public static String uploadFilePath;
  public static int daysToKeepResults;
  public static int daysToShowResults;
  public static String tzTimeZone;
  public static ZoneId zoneId;
  public static String language_country;
  public static Locale locale;
  public static String fiscalYearCode;
  public static String permLocationName;
  public static String permELocationName;
  public static String noteTypeName;
  public static String materialType;
  public static String textForElectronicResources;
  public static boolean importInvoice;
  public static boolean failIfNoInvoiceData;
  public static String paymentMethod;
  public static String permLocationWithInvoiceImport;
  public static String permELocationWithInvoiceImport;
  public static String marcMapping;
  public static boolean exitOnConfigErrors;
  public static boolean exitOnAccessErrors;
  public static boolean exitOnFailedIdLookups;
  public static String onValidationErrors;
  public static String onIsbnInvalid;
  public static boolean onIsbnInvalidFlagError;
  public static boolean onIsbnInvalidRemoveIsbn;
  public static boolean onIsbnInvalidDoNothing;
  public static boolean onValidationErrorsCancelAll;
  public static boolean onValidationErrorsSKipFailed;
  public static boolean onValidationErrorsAttemptImport;
  public static String purchaseOrderUnit;
  public static boolean createOnePurchaseOrderPerFile;
  public static boolean createOnePurchaseOrderPerRecord;

  private static final String EMPTY = "";
  private static final String NOT_APPLICABLE = "NA";
  private static ServletContext context;

  private static final String V_DEFAULT_UI_INVENTORY_PATH = "inventory/view";
  private static final String V_DEFAULT_UI_ORDERS_PATH = "orders/view";
  private static final String V_DEFAULT_MATERIAL_TYPE = "unspecified";
  private static final String V_DEFAULT_FILE_STORAGE_PATH = "/tmp/folio-order-import/";
  private static final int V_DEFAULT_DAYS_TO_KEEP_RESULTS = 365;
  private static final int V_DEFAULT_DAYS_TO_DISPLAY_RESULTS = 14;
  private static final String V_DEFAULT_TIME_ZONE = "Europe/Stockholm";
  private static final String V_DEFAULT_LOCALE = "sv-SE";
  private static final String V_DEFAULT_PURCHASE_ORDER_UNIT = V_PURCHASE_ORDER_UNIT_RECORD;

  public static boolean acquisitionMethodsApiPresent = true;

  @SuppressWarnings("unused") // used in JSP
  public static List<Object> allSettings = new ArrayList<>();

  public static void load (ServletContext servletContext) {
    if (context == null) {
      context = servletContext;
      // FOLIO access
      baseOkapiEndpoint = withEndingSlash(getText(P_BASE_OKAPI_ENDPOINT));
      if (baseOkapiEndpoint.isEmpty()) {
        baseOkapiEndpoint = withEndingSlash(getText(P_BASE_OKAP_ENDPOINT)); //previous spelling
      }
      apiUsername = getText(P_OKAPI_USERNAME);
      apiPassword = getText(P_OKAPI_PASSWORD);
      tenant = getText(P_TENANT);

      // Operations
      uploadFilePath = withEndingSlash(getText(P_UPLOAD_FILE_PATH, V_DEFAULT_FILE_STORAGE_PATH));
      daysToKeepResults = getInt(P_DAYS_TO_KEEP_RESULTS, V_DEFAULT_DAYS_TO_KEEP_RESULTS);

      // MARC mapping
      marcMapping = getText(P_MARC_MAPPING, V_MARC_MAPPING_CHI).toLowerCase();

      // Default values
      permLocationName = getText(P_PERM_LOCATION); // Default, could change with invoice
      permELocationName = getText(P_PERM_E_LOCATION); // Default, could change with invoice
      fiscalYearCode = getText(P_FISCAL_YEAR_CODE);
      noteTypeName = getText(P_NOTE_TYPE);
      materialType = getText(P_MATERIAL_TYPE, V_DEFAULT_MATERIAL_TYPE);
      paymentMethod = getText(P_PAYMENT_METHOD);
      textForElectronicResources = getText(P_TEXT_FOR_ELECTRONIC_RESOURCES);

      // Processing instructions
      purchaseOrderUnit = getText(P_PURCHASE_ORDER_UNIT, V_DEFAULT_PURCHASE_ORDER_UNIT).toLowerCase();
      createOnePurchaseOrderPerFile = purchaseOrderUnit.equals(V_PURCHASE_ORDER_UNIT_FILE);
      createOnePurchaseOrderPerRecord = purchaseOrderUnit.equals(V_PURCHASE_ORDER_UNIT_RECORD);
      exitOnConfigErrors = getBoolean(P_EXIT_ON_CONFIG_ERRORS,true);
      exitOnAccessErrors = getBoolean(P_EXIT_ON_ACCESS_ERRORS, true);
      exitOnFailedIdLookups = getBoolean(P_EXIT_ON_FAILED_ID_LOOKUPS,true);
      importInvoice = getBoolean(P_IMPORT_INVOICE,false);
      failIfNoInvoiceData = getBoolean(P_FAIL_IF_NO_INVOICE_DATA, importInvoice);
      onValidationErrors = getText(P_ON_VALIDATION_ERRORS, V_ON_VALIDATION_ERRORS_CANCEL_ALL);
      onValidationErrorsCancelAll = V_ON_VALIDATION_ERRORS_CANCEL_ALL.equalsIgnoreCase(onValidationErrors);
      onValidationErrorsAttemptImport = V_ON_VALIDATION_ERRORS_ATTEMPT_IMPORT.equalsIgnoreCase(onValidationErrors);
      onValidationErrorsSKipFailed = V_ON_VALIDATION_ERRORS_SKIP_FAILED.equalsIgnoreCase(onValidationErrors);
      onIsbnInvalid = getText(P_ON_ISBN_INVALID, V_ON_ISBN_INVALID_REPORT_ERROR);
      onIsbnInvalidFlagError = V_ON_ISBN_INVALID_REPORT_ERROR.equalsIgnoreCase(onIsbnInvalid);
      onIsbnInvalidRemoveIsbn = V_ON_ISBN_INVALID_REMOVE_ISBN.equalsIgnoreCase(onIsbnInvalid);
      onIsbnInvalidDoNothing = V_ON_ISBN_INVALID_DO_NOTHING.equalsIgnoreCase(onIsbnInvalid);

      // UI
      daysToShowResults = getInt(P_DAYS_TO_SHOW_RESULTS, V_DEFAULT_DAYS_TO_DISPLAY_RESULTS);
      folioUiUrl = withEndingSlash(getText(P_FOLIO_UI_URL));
      folioUiInventoryPath = withEndingSlash(getText(P_FOLIO_UI_INVENTORY_PATH,V_DEFAULT_UI_INVENTORY_PATH));
      folioUiOrdersPath = withEndingSlash(getText(P_FOLIO_UI_ORDERS_PATH, V_DEFAULT_UI_ORDERS_PATH));
      tzTimeZone = getText(P_TZ_TIME_ZONE, V_DEFAULT_TIME_ZONE);
      language_country = getText(P_LOCALE, V_DEFAULT_LOCALE);
      locale = Locale.forLanguageTag(getText(P_LOCALE, V_DEFAULT_LOCALE));
      zoneId = getZoneId();

      permLocationWithInvoiceImport = getText(P_PERM_LOCATION_WITH_INVOICE_IMPORT,
              (importInvoice ? EMPTY : NOT_APPLICABLE));
      permELocationWithInvoiceImport = getText(P_PERM_E_LOCATION_WITH_INVOICE_IMPORT,
              (importInvoice ? EMPTY : NOT_APPLICABLE));
    }
  }

  private static final String POM_PROPERTIES_FILE = "properties-from-pom.properties";

  @SuppressWarnings("unused") // Used in JSP
  public static String getOitVersion() {
    try {
      PropertiesReader reader = new PropertiesReader(POM_PROPERTIES_FILE);
      return reader.getProperty("name") + ", version " + reader.getProperty("version");
    } catch (IOException ioe) {
      return "Could not retrieve the version of this Order Import Tool: " + ioe.getMessage();
    } catch (NullPointerException npe) {
      return "Could not retrieve the version of this Order Import Tool from " + POM_PROPERTIES_FILE;
    }
  }

  public static String getText (String key, String defaultStr) {
    String prop = (String) context.getAttribute(key);
    if (prop == null || prop.isEmpty()) {
      return defaultStr;
    } else {
      return prop;
    }
  }

  public static String getText (String key) {
    String prop = (String) context.getAttribute(key);
    return prop == null ? EMPTY : prop;
  }

  public static boolean getBoolean (String key, boolean defaultsTo) {
    List<String> trueStrings = Arrays.asList("true","TRUE","1","yes","YES","Y");
    List<String> falseStrings = Arrays.asList("false","FALSE","0","NO","no");
    String prop = (String) context.getAttribute(key);
    if (trueStrings.contains(prop)) {
      return true;
    } else if (falseStrings.contains(prop)) {
      return false;
    } else {
      return defaultsTo;
    }
  }

  public static int getInt (String key, int defaultsTo) {
    if (key == null || key.isEmpty()) {
      return defaultsTo;
    } else {
      try {
        return Integer.parseInt((String) context.getAttribute(key));
      }
      catch (NumberFormatException nfe) {
        return defaultsTo;
      }
    }
  }

  /**
   * Ensure ending slash
   * @param str string to check for ending slash
   * @return string with ending slash
   */
  public static String withEndingSlash(String str) {
    if (str == null || str.isEmpty()) {
      return str;
    } else {
      return str.endsWith("/") ? str : str + "/";
    }
  }

  /**
   * Retrive ID of the time zone
   * @return time zone ID
   */
  public static ZoneId getZoneId () {
    try {
      return ZoneId.of(getText(P_TZ_TIME_ZONE, V_DEFAULT_TIME_ZONE));
    } catch (ZoneRulesException zre) {
      return ZoneId.systemDefault();
    }
  }

  /**
   * Retrieve a given MARC-to-FOLIO mapping for a record, given the mapping selected in configuration.
   * @param record MARC record to map to FOLIO
   * @return selected mapping to FOLIO
   */
  public static MarcToFolio getMarcMapping (Record record) {
    switch (marcMapping) {
      case V_MARC_MAPPING_LAMBDA:
        return new MarcMapLambda(record);
      case V_MARC_MAPPING_SIGMA:
        return new MarcMapSigma(record);
      default:
        return new MarcMapChi(record);
    }
  }

}
